src/Versioned.php

```diff
@@ -2681,6 +2681,30 @@ public function writeToStage($stage, $forceInsert = false)
             }
         });
     }
+    
+    /**
+     * write to draft and if a published version already exists and it is currently the same as draft then 
+     * publish the change as well.
+     * 
+     * This allows you to update pages and keep them in the same state (published / modified / draft) as before
+     */
+    public function writeAndPublishIfAppropriate(?bool $recursive = true, ?bool $forceInsert = false)
+    {
+        // check reading mode
+        $myStage = Versioned::get_stage();
+        Versioned::set_stage(Versioned::DRAFT);
+        // is it on live and is live the same as draft
+        $canBePublished = $this->owner->isPublished() && ! $this->owner->isModifiedOnDraft();
+        $this->owner->writeToStage(Versioned::DRAFT, $forceInsert);
+        if ($canBePublished) {
+            if ($recursive) {
+                $this->owner->publishSingle();
+            } else {
+                $this->owner->publishRecursive();
+            }
+        }    
+        Versioned::set_stage($myStage);
```

Comment:
You don't need to save and restore the old stage, you can use `withVersionedMode` and after the callback is processed the current stage is restored automatically.

Also the `publishSingle()` and `publishRecursive()` calls seemed to be the other way round in relation to the `$recursive` condition.

```suggestion
        Versioned::withVersionedMode(function () use ($recursive, $forceInsert) {
            Versioned::set_stage(Versioned::DRAFT);
	        // is it on live and is live the same as draft
	        $canBePublished = $this->owner->isPublished() && ! $this->owner->isModifiedOnDraft();
	        $this->owner->writeToStage(Versioned::DRAFT, $forceInsert);
	        if ($canBePublished) {
	            if ($recursive) {
	                $this->owner->publishRecursive();
	            } else {
	                $this->owner->publishSingle();
	            }
	        }    
        }
```

Reply:
> withVersionedMode

I prefer not to use callbacks, as this feels easier to debug / read.  Would you be ok with that? 

> recursive swap

oooops!  sorry, fixed

Reply:
> > withVersionedMode
> 
> I prefer not to use callbacks, as this feels easier to debug / read. Would you be ok with that?

Taking a quick look the two methods right above this suggested addition (and likely a lot of others in the class) are using the `static::withVersionedMode` callback approach, so I would prefer to keep that for consistency.

Reply:
I agree that `withVersionedMode` should be used